guides-markdown/bcoin-clock.md

```diff
@@ -0,0 +1,560 @@
+
+```javascript
+// Get the SPV node object from the globally-installed bcoin package
+const {SPVNode} = require('bcoin');
+
+// Configure the node for mainnet, write logs, use the database on disk, etc
+const node = new SPVNode({
+  network: 'main',
+  config: true,
+  argv: true,
+  env: true,
+  logFile: true,
+  logConsole: true,
+  logLevel: 'debug',
+  db: 'leveldb',
+  memory: false,
+  persistent: true,
+  workers: true,
+  listen: true,
+  loader: require
+});
+
+(async () => {
+
+  // Validate the prefix directory (probably ~/.bcoin)
+  await node.ensure();
+  // Open the node and all its child objects, wait for the database to load
+  await node.open();
+  // Connect to the network
+  await node.connect();
+
+  // TODO: add cool stuff here in the next step! (*)
+
+  // Start the blockchain sync
+  node.startSync();
+  
+})().catch((err) => {
+  console.error(err.stack);
+  process.exit(1);
+});
+```
+
+### Respond to events
+
+The line marked `(*)` above is where we can ask bcoin to react to certain "events" like a new block being added to the tip of the blockchain, or a transaction being received that affects our own wallet. For our purposes, we want bcoin to write a JSON file for each new block, so we can display the block details on the clock! First let's add a function (at the very top of the file) that writes these files, labeled by the block's height. We only need to keep 20 or so of the most recent blocks, so we'll allow the script to prune the directory as well. On a Raspberry Pi, the script below will create the directory at `/home/pi/blocks` (or `~/blocks`),
+
+```javascript
+// allow the script to access the file system on disk
+const fs = require('fs');
+const os = require('os');
+const maxFiles = 20;
+
+// create the a directory for block headers if it doesn't exist already
+const blocksDir = os.homedir() + '/blocks/';
+if (!fs.existsSync(blocksDir)){
+    fs.mkdirSync(blocksDir);
+}
+
+// Write a JSON object to disk
+function writeFile(index, element, dir){
+	fs.writeFile(dir + index, JSON.stringify(element), function(err){
+		if(!err){
+			// delete files older than 20 blocks if it exists
+			try {
+				fs.unlinkSync(dir + (index-maxFiles));
+			} catch(err) {
+				console.log(err);
+			}
+		}
+	});
+}
+```
+
+Then we can tell bcoin to call the above function whenever a new block arrives. This code will go right where the `(*)` line is in our script:
+
+```javascript
+  // write new block details to file
+  node.on('block', async (block) => {
+  	// most of the block's details are returned by the 'block' event but we need to get its height from the blockchain database
+  	headers = await node.chain.getEntryByHash(block.hash('hex'));
+  	blockHeight = headers.height;
+
+	// simplify the block data structure to a string
```

Comment:
"Make the block data human readable" might be more accurate. Basically toJSON decodes the raw data which I believe is in hex when it comes back from the server.